### DISCUSSION TITLE
How can I mock a function that the function being tested uses?

### DISCUSSION THREAD

**Discussion by paulwongx** [vitest-dev/vitest#3548] [Q&A] [answered]
For example, how can I mock the `hello` function and `.mockReturnValue("Hi")`? Hoping someone can assist with this simple example as it's quite confusing to me currently.

```js
// greet.ts
export function hello() {
  return "Hello ";
}

export function greet(name: string) {
  return hello() + name;
}
```

Test file:
```js
// greet.spec.ts
import {hello, greet} from "./greet";

vi.mock("./greet", async () => {
  const mod = await vi.importActual<typeof import("./greet")>("./greet");
  const hello = vi.fn().mockReturnValue("Hi ");
  return {
    ...mod,
    hello,
    // greet: vi.fn((name: string)=> hello() + name),
    // ^this works if I uncomment it, but it defeats 
    // the purpose of the test. 
    // Is it just not possible to do what I want? 
  }
})

describe("Greeting", () => {
  test("should return a personalized greeting", async () => {
    const result = greet("John");

    expect(result).toBe("Hi John");
    expect(hello).toHaveBeenCalled();
  });
});
```
Running the test still gives `Hello John` and the `hello` function is not called. No surprise... how can I mock the implementation of `hello()` in the actual `greet` function?

**Comment by sheremet-va** [comment, marked as answer]
You cannot.

**Reply by burtek** [reply]
You can if the `hello` function is defined in another module.

**Reply by tushar-sg** [reply]
could you please expand on the answer. I faced this issue at work multiple times and thought maybe I am doing something wrong. 

If a file has two exported functions `a` and `b`, `a` uses `b` internally. Can we not call `a` and mock `b`? 

If not, what's the recommended approach to test such scenarios?

Basically this issue 👉🏻 https://github.com/vitest-dev/vitest/discussions/3667

**Reply by sheremet-va** [reply]
Spying only spies calls made on an object by redefining the method with another method that tracks all calls:

```
obj.exportedMethod()
```

This is how these methods are called when imported. But when you call it inside a module, you call the original reference that cannot be intercepted.

**Reply by tushar-sg** [reply]
That makes sense! Thanks

**Reply by bkhouri** [reply]
This is a test-only problem, and having to re-organized the production code to make the tests "work" seems like a framework/language limitation.  Is it possible to create a plugin, or something, that would "mock" all instances of a function?

it's times like this where I miss working with `pytest`'s [`monkeypatch` fixture](https://docs.pytest.org/en/7.4.x/reference/reference.html#std-fixture-monkeypatch).

**Reply by sheremet-va** [reply]
I don’t believe this is a test framework problem. If language doesn’t allow something, then don’t write code that doesn’t work - it’s that simple. 

Write your code to be testable. Documentation has a section about it - https://vitest.dev/guide/mocking.html#mocking-pitfalls

**Reply by bkhouri** [reply]
Well, it doesn't make sense to move an internal method to its own file, and export it, simply for the test!
